lib/lints/banned_external_code/banned_external_code_rule.dart

```diff
@@ -0,0 +1,315 @@
+        rootPath: rootPath,
+      )) {
+        continue;
+      }
+
+      final entryCode = super.code.copyWith(
+            errorSeverity: entry.severity ??
+                parameters.severity ??
+                super.code.errorSeverity,
+            problemMessage: entry.reason,
+          );
+
+      switch (entry) {
+        // all non null
+        case BannedExternalCodeEntryParameters(
+            :final id?,
+            :final className?,
+            :final source?
+          ):
+          linter.banIdFromClassFromSource(entryCode, id, className, source);
+
+        // two non nulls
+        case BannedExternalCodeEntryParameters(
+            :final className?,
+            :final source?
+          ):
+          linter.banClassFromSource(entryCode, className, source);
+        case BannedExternalCodeEntryParameters(:final id?, :final className?):
+          linter.banIdFromClass(entryCode, id, className);
+        case BannedExternalCodeEntryParameters(:final id?, :final source?):
+          linter.banIdFromSource(entryCode, id, source);
+
+        // one non null
+        case BannedExternalCodeEntryParameters(:final source?):
+          linter.banSource(entryCode, source);
+        case BannedExternalCodeEntryParameters(:final className?):
+          linter.banClass(entryCode, className);
+        case BannedExternalCodeEntryParameters(:final id?):
+          linter.banId(entryCode, id);
+
+        // all null
+        case BannedExternalCodeEntryParameters():
+          break;
+      }
+    }
+  }
+}
+
+class _BannedCodeLinter {
+  const _BannedCodeLinter({
+    required this.resolver,
+    required this.reporter,
+    required this.context,
+    required this.config,
+  });
+
+  final CustomLintResolver resolver;
+  final ErrorReporter reporter;
+  final CustomLintContext context;
+  final RuleConfig<BannedExternalCodeParameters?> config;
+
+  bool _matchesSource(String? parentSourcePath, String sourceToMatch) {
+    if (parentSourcePath == null) {
+      return false;
+    }
+
+    final libraryPathParts = parentSourcePath.split('/');
+    final libraryName = libraryPathParts.first;
+
+    final sourceLibraryPathParts = sourceToMatch.split('/');
+    final sourceLibraryName = sourceLibraryPathParts.first;
+
+    final matchesLibraryName = libraryName == sourceLibraryName;
+    final shouldLintFileFromSource = sourceLibraryPathParts.length > 1;
+    if (shouldLintFileFromSource) {
+      final sourceFile = (sourceLibraryPathParts..removeAt(0)).join('/');
+      final libraryFile = (libraryPathParts..removeAt(0)).join('/');
+
+      final matchesFile = p.equals(sourceFile, libraryFile);
+      return matchesLibraryName && matchesFile;
+    }
+
+    return matchesLibraryName;
+  }
+
+  void banId(
+    LintCode entryCode,
+    String id,
+  ) {
+    banIdFromSource(entryCode, id);
+  }
+
+  void banClass(
+    LintCode entryCode,
+    String className,
+  ) {
+    banClassFromSource(entryCode, className);
+  }
```

Comment:
Not sure these trampoline methods are useful. Maybe we should call the underlying function directly?

Reply:
Yeah I agree is it a little weird, it's a holdover from before I refactored it to remove the duplicate code. I decided to keep it for the sake of consistency/readability.

If we make source required, these two can be removed entirely.